action/governance/withdrawFunds.go

```diff
@@ -0,0 +1,234 @@
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "withdraw_proposal_does_not_meet_withdraw_requirement"),
+		}
+		return false, result
+	}
+	// 2. change outcome, status, state
+	proposal.Outcome = governance.ProposalOutcomeInsufficientFunds
+	proposal.Status = governance.ProposalStatusCompleted
+	err = ctx.ProposalMasterStore.Proposal.WithPrefixType(governance.ProposalStateFailed).Set(proposal)
+	if err != nil {
+		ctx.Logger.Error("Failed to add proposal to FAILED store :", proposal.ProposalID)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "failed_to_add_proposal_to_failed_store"),
+		}
+		return false, result
+	}
+	ok, err := ctx.ProposalMasterStore.Proposal.WithPrefixType(governance.ProposalStateActive).Delete(proposal.ProposalID)
+	if err != nil || !ok {
+		ctx.Logger.Error("Failed to delete proposal from ACTIVE store :", proposal.ProposalID)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "failed_to_delete_proposal_from_active_store"),
+		}
+		return false, result
+	}
+
+	// 3. Check if the contributor has funded this proposal, if so, get the amount of funds
+	proposalFund, err := governance.GetCurrentFundsByContributor(proposal.ProposalID, withdrawProposal.Contributor, ctx.ProposalMasterStore.ProposalFund)
+	if err != nil {
+		ctx.Logger.Error("No available funds to withdraw for this contributor :", withdrawProposal.Contributor)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "no_available__fund_to_withdraw_for_this_contributor"),
+		}
+		return false, result
+	}
+
+	// 4. Check if the contributor has sufficient funds to withdraw for that proposal
+	withdrawAmount := balance.NewAmountFromBigInt(withdrawProposal.WithdrawValue.Value.BigInt())
+	if proposalFund.BigInt().Cmp(withdrawAmount.BigInt()) < 0 {
+		ctx.Logger.Error("Insufficient funds to withdraw for this contributor :", withdrawProposal.Contributor)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "no_available__fund_to_withdraw_for_this_contributor"),
+			Log: action.ErrNotEnoughFund.Msg,
+		}
+		return false, result
+	}
+
+	// 5. withdraw
+	// deduct from proposal fund
+	err = ctx.ProposalMasterStore.ProposalFund.DeductFunds(proposal.ProposalID, withdrawProposal.Contributor, withdrawAmount)
+	if err != nil {
+		ctx.Logger.Error("Failed to deduct funds from proposal:", withdrawProposal.ProposalID)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "withdraw_proposal_deduct_fund_failed"),
+		}
+		return false, result
+	}
+	// add to beneficiary address
+	coin := withdrawProposal.WithdrawValue.ToCoin(ctx.Currencies)
+	err = ctx.Balances.AddToAddress(withdrawProposal.Beneficiary.Bytes(), coin)
+	if err != nil {
+		// return funds to proposal
+		err = ctx.ProposalMasterStore.ProposalFund.AddFunds(proposal.ProposalID, withdrawProposal.Contributor, withdrawAmount)
```

Comment:
Can you check with Kevin/Tanmay if we need rollback in Tx if something like 'AddToAddress/Minus' this happened? I think this rollback could be already handled somewhere else, because what if this AddFunds rollback fails again and we still lost the money.

Reply:
yeah let me ask. and here I implemented this because there is same logic in createProposal

Reply:
as we spoke, added panic inside